File: 哈希/LC 49字母异位词分组.cpp
#include<bits/stdc++.h>
#include<vector>
#include<algorithm>
#include<ranges>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        //跑一遍，我的想法是异或结果相同。
        unordered_map<string,vector<string>> hashMap;
        for(string& s: strs){
            string sorted_s = s;
            ranges::sort(sorted_s);
            hashMap[sorted_s].push_back(s);
        }

        vector<vector<string>> ans;
        ans.reserve(hashMap.size());
        for (auto& [_,value] : hashMap){
            ans.push_back(value);
        }
        return ans;
    }
};

File: 哈希/LC 49字母异位词分组.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> hashMap = new HashMap<>();
        for (String s: strs){
            char[] sortedS = s.toCharArray();
            Arrays.sort(sortedS);
            // computeIfAbsent：如果 key 不在哈希表中，则插入一个新的 ArrayList
            hashMap.computeIfAbsent(
                new String(sortedS),_->new ArrayList<>()
                ).add(s);
        }
        return new ArrayList<>(hashMap.values());
    }
}